File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunitySearchFragment.java
package com.veganbeauty.app.features.community.com_feed;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.ConcatAdapter;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.veganbeauty.app.R;
import com.veganbeauty.app.core.base.RootieFragment;
import com.veganbeauty.app.data.local.LocalJsonReader;
import com.veganbeauty.app.data.local.RootieDatabase;
import com.veganbeauty.app.data.local.entities.CommunityPostEntity;
import com.veganbeauty.app.data.local.entities.CommunityProduct;
import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.data.local.entities.ReelEntity;
import com.veganbeauty.app.data.local.entities.UserEntity;
import com.veganbeauty.app.data.remote.FirestoreService;
import com.veganbeauty.app.data.repository.CommunityRepository;
import com.veganbeauty.app.databinding.ComFragmentSearchBinding;
import com.veganbeauty.app.features.community.profile.CommunityProfileFragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class CommunitySearchFragment extends RootieFragment {

    private ComFragmentSearchBinding binding;

    private CommunityViewModel viewModel;
    private UserSearchAdapter userAdapter;
    private PostAdapter postAdapter;
    private ConcatAdapter concatAdapter;
    
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = ComFragmentSearchBinding.inflate(inflater, container, false);
        setupViewModel();
        return binding.getRoot();
    }

    private void setupViewModel() {
        RootieDatabase db = RootieDatabase.getDatabase(requireContext());
        CommunityRepository repository = new CommunityRepository(db.communityDao(), new LocalJsonReader(requireContext()), new FirestoreService());
        CommunityViewModelFactory factory = new CommunityViewModelFactory(repository);
        viewModel = new ViewModelProvider(requireActivity(), factory).get(CommunityViewModel.class);
    }

    @Override
    public void setupUI(@NonNull View view) {
        binding.ivBack.setOnClickListener(v -> getParentFragmentManager().popBackStack());

        binding.ivClear.setOnClickListener(v -> binding.etSearch.setText(""));

        userAdapter = new UserSearchAdapter(new ArrayList<>(), user -> {
            CommunityProfileFragment fragment = CommunityProfileFragment.newInstance(user.getUser_id());
            getParentFragmentManager().beginTransaction()
                    .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out, android.R.anim.fade_in, android.R.anim.fade_out)
                    .replace(R.id.main_container, fragment)
                    .addToBackStack(null)
                    .commit();
        });

        postAdapter = new PostAdapter();

        concatAdapter = new ConcatAdapter(userAdapter, postAdapter);
        binding.rvSearchResults.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.rvSearchResults.setAdapter(concatAdapter);

        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String query = s != null ? s.toString().trim() : "";
                binding.ivClear.setVisibility(query.isEmpty() ? View.GONE : View.VISIBLE);
                binding.llSuggestions.setVisibility(query.isEmpty() ? View.VISIBLE : View.GONE);
                binding.rvSearchResults.setVisibility(query.isEmpty() ? View.GONE : View.VISIBLE);
                performSearch(query);
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });

        for (int i = 0; i < binding.llSuggestionItems.getChildCount(); i++) {
            View child = binding.llSuggestionItems.getChildAt(i);
            if (child instanceof TextView) {
                TextView textView = (TextView) child;
                textView.setOnClickListener(v -> {
                    String text = textView.getText().toString().replace("• ", "").trim();
                    binding.etSearch.setText(text);
                    binding.etSearch.setSelection(text.length());

                    InputMethodManager imm = (InputMethodManager) requireContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                    if (imm != null) {
                        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                    }
                });
            }
        }

        binding.llRefresh.setOnClickListener(v -> {
            List<View> children = new ArrayList<>();
            for (int i = 0; i < binding.llSuggestionItems.getChildCount(); i++) {
                children.add(binding.llSuggestionItems.getChildAt(i));
            }
            Collections.shuffle(children);
            binding.llSuggestionItems.removeAllViews();
            for (View child : children) {
                binding.llSuggestionItems.addView(child);
            }
        });

        binding.tvSearchAction.setOnClickListener(v -> {
            InputMethodManager imm = (InputMethodManager) requireContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            if (imm != null) {
                imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
            }
        });

        binding.etSearch.requestFocus();
        InputMethodManager imm = (InputMethodManager) requireContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.showSoftInput(binding.etSearch, InputMethodManager.SHOW_IMPLICIT);
        }
    }

    private void performSearch(String query) {
        if (query.isEmpty()) {
            userAdapter.updateData(new ArrayList<>());
            postAdapter.updateData(new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
            binding.tvEmpty.setVisibility(View.GONE);
            return;
        }

        String lowerQuery = query.toLowerCase();

        List<UserEntity> allUsers = viewModel.getUsers().getValue() != null ? viewModel.getUsers().getValue() : new ArrayList<>();
        
        List<CommunityPostEntity> allPosts = viewModel.getPosts().getValue() != null ? viewModel.getPosts().getValue() : new ArrayList<>();
        
        final List<ReelEntity> allReels = viewModel.getReels().getValue() != null ? viewModel.getReels().getValue() : new ArrayList<>();

        List<UserEntity> matchedUsers = new ArrayList<>();
        for (UserEntity user : allUsers) {
            if ((user.getUsername() != null && user.getUsername().toLowerCase().contains(lowerQuery)) ||
                (user.getFull_name() != null && user.getFull_name().toLowerCase().contains(lowerQuery))) {
                matchedUsers.add(user);
            }
        }

        List<CommunityPostEntity> matchedPosts = new ArrayList<>();
        for (CommunityPostEntity post : allPosts) {
            boolean matchesContent = post.getContent() != null && post.getContent().toLowerCase().contains(lowerQuery);
            boolean matchesType = post.getType() != null && post.getType().toLowerCase().contains(lowerQuery);
            boolean matchesSkinType = post.getSkinType() != null && post.getSkinType().toLowerCase().contains(lowerQuery);
            boolean matchesConcern = post.getConcern() != null && post.getConcern().toLowerCase().contains(lowerQuery);
            
            boolean matchesAuthor = false;
            for (UserEntity u : matchedUsers) {
                if (u.getUser_id() != null && u.getUser_id().equals(post.getAuthorId())) {
                    matchesAuthor = true;
                    break;
                }
            }
            
            if (matchesContent || matchesType || matchesSkinType || matchesConcern || matchesAuthor) {
                matchedPosts.add(post);
            }
        }

        executor.execute(() -> {
            Context ctx = getContext();
            if (ctx == null) return;
            
            final List<CommunityProduct> finalProductsList = new LocalJsonReader(ctx).getProducts();
            
            requireActivity().runOnUiThread(() -> {
                userAdapter.updateData(matchedUsers);
                postAdapter.updateData(matchedPosts, new ArrayList<>(), allReels, finalProductsList);

                if (matchedUsers.isEmpty() && matchedPosts.isEmpty()) {
                    binding.tvEmpty.setVisibility(View.VISIBLE);
                } else {
                    binding.tvEmpty.setVisibility(View.GONE);
                }
            });
        });
    }

    @Override
    protected void observeViewModel() {
        // Not used
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        InputMethodManager imm = (InputMethodManager) requireContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null && getView() != null) {
            imm.hideSoftInputFromWindow(getView().getWindowToken(), 0);
        }
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/community/com_feed/CommunitySearchFragment.kt
package com.veganbeauty.app.features.community.com_feed

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ConcatAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.room.Room
import com.veganbeauty.app.R
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.data.local.LocalJsonReader
import com.veganbeauty.app.data.local.RootieDatabase
import com.veganbeauty.app.data.remote.FirestoreService
import com.veganbeauty.app.data.repository.CommunityRepository
import com.veganbeauty.app.databinding.ComFragmentSearchBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class CommunitySearchFragment : RootieFragment() {

    private var _binding: ComFragmentSearchBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: CommunityViewModel
    private lateinit var userAdapter: UserSearchAdapter
    private lateinit var postAdapter: PostAdapter
    private lateinit var concatAdapter: ConcatAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = ComFragmentSearchBinding.inflate(inflater, container, false)
        setupViewModel()
        return binding.root
    }

    private fun setupViewModel() {
        val db = RootieDatabase.getDatabase(requireContext())
        val repository = CommunityRepository(db.communityDao(), LocalJsonReader(requireContext()), FirestoreService())
        val factory = CommunityViewModelFactory(repository)
        viewModel = ViewModelProvider(requireActivity(), factory)[CommunityViewModel::class.java]
    }

    override fun setupUI(view: View) {
        binding.ivBack.setOnClickListener {
            parentFragmentManager.popBackStack()
        }

        binding.ivClear.setOnClickListener {
            binding.etSearch.setText("")
        }

        userAdapter = UserSearchAdapter(emptyList()) { user ->
            // Open Profile
            val fragment = com.veganbeauty.app.features.community.profile.CommunityProfileFragment.newInstance(user.user_id)
            parentFragmentManager.beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out, android.R.anim.fade_in, android.R.anim.fade_out)
                .replace(R.id.main_container, fragment)
                .addToBackStack(null)
                .commit()
        }

        postAdapter = PostAdapter()

        concatAdapter = ConcatAdapter(userAdapter, postAdapter)
        binding.rvSearchResults.layoutManager = LinearLayoutManager(requireContext())
        binding.rvSearchResults.adapter = concatAdapter

        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val query = s?.toString()?.trim() ?: ""
                binding.ivClear.visibility = if (query.isNotEmpty()) View.VISIBLE else View.GONE
                binding.llSuggestions.visibility = if (query.isEmpty()) View.VISIBLE else View.GONE
                binding.rvSearchResults.visibility = if (query.isEmpty()) View.GONE else View.VISIBLE
                performSearch(query)
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        // Setup suggestions
        for (i in 0 until binding.llSuggestionItems.childCount) {
            val textView = binding.llSuggestionItems.getChildAt(i) as? android.widget.TextView
            textView?.setOnClickListener {
                val text = textView.text.toString().replace("• ", "").trim()
                binding.etSearch.setText(text)
                binding.etSearch.setSelection(text.length)
                
                // Hide keyboard to show results
                val imm = requireContext().getSystemService(android.content.Context.INPUT_METHOD_SERVICE) as android.view.inputmethod.InputMethodManager
                imm.hideSoftInputFromWindow(view.windowToken, 0)
            }
        }

        binding.llRefresh.setOnClickListener {
            val children = mutableListOf<View>()
            for (i in 0 until binding.llSuggestionItems.childCount) {
                children.add(binding.llSuggestionItems.getChildAt(i))
            }
            children.shuffle()
            binding.llSuggestionItems.removeAllViews()
            children.forEach { binding.llSuggestionItems.addView(it) }
        }

        binding.tvSearchAction.setOnClickListener {
            val imm = requireContext().getSystemService(android.content.Context.INPUT_METHOD_SERVICE) as android.view.inputmethod.InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }

        // Request focus
        binding.etSearch.requestFocus()
        val imm = requireContext().getSystemService(android.content.Context.INPUT_METHOD_SERVICE) as android.view.inputmethod.InputMethodManager
        imm.showSoftInput(binding.etSearch, android.view.inputmethod.InputMethodManager.SHOW_IMPLICIT)
    }

    private fun performSearch(query: String) {
        if (query.isEmpty()) {
            userAdapter.updateData(emptyList())
            postAdapter.updateData(emptyList(), emptyList(), emptyList(), emptyList())
            binding.tvEmpty.visibility = View.GONE
            return
        }

        val lowerQuery = query.lowercase()

        val allUsers = viewModel.users.value ?: emptyList()
        val allPosts = viewModel.posts.value ?: emptyList()
        val allReels = viewModel.reels.value ?: emptyList()

        val matchedUsers = allUsers.filter {
            it.username.lowercase().contains(lowerQuery) || it.full_name.lowercase().contains(lowerQuery)
        }

        val matchedPosts = allPosts.filter {
            (it.content.lowercase().contains(lowerQuery)) || 
            (it.type?.lowercase()?.contains(lowerQuery) == true) || 
            (it.skinType?.lowercase()?.contains(lowerQuery) == true) || 
            (it.concern?.lowercase()?.contains(lowerQuery) == true) ||
            matchedUsers.any { u -> u.user_id == it.authorId } // also match if author is in matched users
        }

        viewLifecycleOwner.lifecycleScope.launch {
            val ctx = context ?: return@launch
            val productsList = FeedDataCache.productsList ?: withContext(Dispatchers.IO) { LocalJsonReader(ctx).getProducts() }.also { FeedDataCache.productsList = it }

            userAdapter.updateData(matchedUsers)
            postAdapter.updateData(matchedPosts, emptyList(), allReels, productsList)

            if (matchedUsers.isEmpty() && matchedPosts.isEmpty()) {
                binding.tvEmpty.visibility = View.VISIBLE
            } else {
                binding.tvEmpty.visibility = View.GONE
            }
        }
    }

    override fun observeViewModel() {
        // Observers are not strictly needed because we only search over loaded data
    }

    override fun onDestroyView() {
        super.onDestroyView()
        val imm = requireContext().getSystemService(android.content.Context.INPUT_METHOD_SERVICE) as android.view.inputmethod.InputMethodManager
        imm.hideSoftInputFromWindow(view?.windowToken, 0)
        _binding = null
    }
}
